pkg/auth/authmap_gc.go

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/cilium/cilium/pkg/identity"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/node/addressing"
+)
+
+type authMapGarbageCollector struct {
+	authmap authMap
+	ipCache ipCache
+
+	discoveredCiliumNodeIDs    map[uint16]struct{}
+	discoveredCiliumIdentities map[identity.NumericIdentity]struct{}
+}
+
+func newAuthMapGC(authmap authMap, ipCache ipCache) *authMapGarbageCollector {
+	return &authMapGarbageCollector{
+		authmap: authmap,
+		ipCache: ipCache,
+		discoveredCiliumNodeIDs: map[uint16]struct{}{
+			0: {}, // Local node 0 is always available
+		},
+		discoveredCiliumIdentities: map[identity.NumericIdentity]struct{}{},
+	}
+}
+
+func (r *authMapGarbageCollector) HandleCiliumNodeEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumNode]) error {
+	switch e.Kind {
+	case resource.Upsert:
+		if r.discoveredCiliumNodeIDs != nil {
+			log.Debug("auth: nodes discovered - getting node id")
+			remoteNodeIDs := r.remoteNodeIDs(e.Object)
+			for _, rID := range remoteNodeIDs {
+				r.discoveredCiliumNodeIDs[rID] = struct{}{}
+			}
+		}
+	case resource.Sync:
+		log.Debug("auth: nodes synced - cleaning up missing nodes")
+		if err := r.cleanupMissingNodes(); err != nil {
+			e.Done(err)
+			return fmt.Errorf("failed to cleanup missing nodes: %w", err)
+		}
+		r.discoveredCiliumNodeIDs = nil
+	case resource.Delete:
+		log.Debugf("auth: node deleted - cleaning up: %s", e.Key.Name)
+		if err := r.cleanupDeletedNode(e.Object); err != nil {
+			e.Done(err)
+			return fmt.Errorf("failed to cleanup deleted node: %w", err)
+		}
+	}
+	e.Done(nil)
+	return nil
+}
+
+func (r *authMapGarbageCollector) HandleCiliumIdentityEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumIdentity]) error {
+	switch e.Kind {
+	case resource.Upsert:
+		if r.discoveredCiliumIdentities != nil {
+			log.Debug("auth: identities discovered")
+			id, err := identity.ParseNumericIdentity(e.Object.Name)
+			if err != nil {
+				return fmt.Errorf("failed to parse identity: %w", err)
+			}
```

Comment:
This branch is missing the e.Done call. I recommend to restructure this and the other handler, make a `err` variable at the start of the function, defer the `e.Done(err)` then set `err` in all of your branches, that way you can't ever forget to call `e.Done`

Reply:
Bit of background: `e.Done` marks the event as handled and without the call we'd never process an event for the same key again. `pkg/k8s/resource` guards against this with a finalizer function that will panic if we ever GC an event that wasn't marked done.

Reply:
@dylandreimerink thanks - good catch! refactored it towards your approach!